mava/networks/retention.py

```diff
@@ -0,0 +1,324 @@
+        # Get the shape of the input: batch size and number of timesteps
+        B, T = dones.shape
+
+        # Create the n and m matrices
+        n = jnp.arange(T)[:, jnp.newaxis, ...]
+        m = jnp.arange(T)[jnp.newaxis, ...]
+
+        # Decay based on difference in timestep indices.
+        decay_matrix = (self.decay_kappa ** (n - m)) * (n >= m)
+        # Replace NaN values with 0
+        decay_matrix = jnp.nan_to_num(decay_matrix)
+
+        # Adjust for batch size
+        decay_matrix = jnp.broadcast_to(decay_matrix, (B, T, T))
+
+        return decay_matrix
+
+    def get_xi(self, dones: Array) -> Array:
+        """Computes a decaying matrix 'xi', which decays over time until the first done signal."""
+        # Get done status for each timestep by slicing out the agent dimension
+        timestep_dones = dones[:, :: self.n_agents]
+        B, T = timestep_dones.shape
+
+        # Compute the first done step for each sequence,
+        # or set it to sequence length if no dones exist
+        first_dones = jnp.where(
+            ~jnp.any(timestep_dones, axis=1, keepdims=True),
+            jnp.full((B, 1), T),
+            jnp.argmax(timestep_dones, axis=1, keepdims=True),
+        )
+
+        # Initialize the decaying matrix 'xi'
+        xi = jnp.zeros((B, T, 1))
+        # Fill 'xi' with decaying values up until the first done step
+        for i in range(T):
+            before_first_done = i < first_dones
+            xi_i = (self.decay_kappa ** (i + 1)) * before_first_done
+            xi = xi.at[:, i, :].set(xi_i)
+
+        # Repeat the decay matrix 'xi' for all agents
+        xi = jnp.repeat(xi, self.n_agents, axis=1)
+
+        return xi
+
+
+class MultiScaleRetention(nn.Module):
+    """Multi-scale retention mechanism for Sable."""
+
+    embed_dim: int
+    n_head: int
+    n_agents: int
+    memory_config: DictConfig
+    full_self_retention: bool = False
+    decay_scaling_factor: float = 1.0
+
+    def setup(self) -> None:
+        assert self.embed_dim % self.n_head == 0, "embed_dim must be divisible by n_head"
+        # Head size
+        self.head_size = self.embed_dim // self.n_head
+
+        # Decay kappa for each head
+        self.decay_kappas = 1 - jnp.exp(
+            jnp.linspace(jnp.log(1 / 32), jnp.log(1 / 512), self.n_head)
+        )
+        self.decay_kappas = self.decay_kappas * self.decay_scaling_factor
+
+        # Initialize the weights and group norm
+        self.w_g = self.param(
+            "w_g",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.embed_dim, self.head_size),
+        )
+        self.w_o = self.param(
+            "w_o",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.head_size, self.embed_dim),
+        )
+        self.group_norm = nn.GroupNorm(num_groups=self.n_head)
+
+        # Initialize the retention mechanisms
```

Comment:
```suggestion
        # Initialize the retention mechanisms
```
```suggestion
        # Initialise the retention mechanisms
```

Reply:
Done, but many other files have this as Initialize so worth maybe make a PR one day just to unify the writing